tola/util.py

```diff
@@ -129,63 +128,97 @@ def saveDataToSilo(silo, data, read=-1, user=None):
             lvs.create_date = timezone.now()
             lvs.read_id = read_source_id
 
-        counter = 0
-
-        row = cleanDataObj(row, silo)
-
+        row = clean_data_obj(row, col_type)
         for key, val in row.iteritems():
-            # if key == "" or key is None or key == "silo_id": continue
-            # elif key == "id" or key == "_id": key = "user_assigned_id"
-            # elif key == "edit_date": key = "editted_date"
-            # elif key == "create_date": key = "created_date"
-            # if type(val) == str or type(val) == unicode:
-            #     val = smart_str(val, strings_only=True).strip()
-            # if fieldToType.get(key, 'string') == 'int':
-            #     try:
-            #         val = int(val)
-            #     except ValueError as e:
-            #         continue
-            # if fieldToType.get(key, 'string') == 'double':
-            #     try:
-            #         val = float(val)
-            #     except ValueError as e:
-            #         continue
-
             if not isinstance(key, tuple):
                 if key not in keys:
                     keys.append(key)
                 setattr(lvs, key, val)
 
         counter += 1
-        lvs = calculateFormulaCell(lvs,silo)
+        lvs = calculateFormulaCell(lvs, silo)
         lvs.save()
-    addColsToSilo(silo, keys)
+
+    addColsToSilo(silo, keys, col_type)
     res = {"skipped_rows": skipped_rows, "num_rows": counter}
     return res
 
 
+def get_value_type(val):
+    """
+    Convert/compute the given value to know the type of it
+    :param val: any
+    :return: string('int', 'double', 'string')
+    """
+    try:
+        number = float(val)
+        result = number % 1
+    except ValueError:
+        return 'string'
+    except TypeError:
+        return type(val).__name__
+    else:
+        if result > 0:
+            return 'double'
+        else:
+            return 'int'
+
+
+def get_column_types(rows):
+    """
+    Find the types of the data passed and returns and
+    the data as a list of dict
+    :param rows: list(dict) or silo.CustomDictReader
+    :return: dict, list(dict)
+    """
+    col_type = {}
+    data = []
 
-def cleanDataObj(obj, silo):
+    for row in rows:
+        data.append(row)
+        for key, val in row.items():
+            key = cleanKey(key)
+            c_type = col_type.get(key, None)
+            if c_type:
+                value_type = get_value_type(val)
+                if c_type == 'int' and value_type == 'double':
+                    col_type[key] = value_type
+                elif c_type != 'string' and value_type == 'string':
+                    col_type[key] = value_type
+            else:
+                col_type.update({key: get_value_type(val)})
+
+    return col_type, data
+
+
+def clean_data_obj(obj, col_type={}):
```

Comment:
Since we are basically handling raw user input here either `clean_data_object` or `cleanKey` should validate the input to prevent mongodb or other possible injections. Or are we handling this issue in another function?

Reply:
I updated the `clean_data_obj` function and was checking other parts of the code and I didn't see such kind of validation. I don't think that we currently have it. We should add it but we should create a ticket for that, this PR is big enough.

Reply:
Yes, it should be another ticket. I will have a look at this issue